micrometer-jakarta/src/main/java/io/micrometer/jakarta/instrument/binder/http/ObservationHttpJakartaInterceptor.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright 2023 VMware, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.jakarta.instrument.binder.http;
+
+import io.micrometer.observation.Observation;
+import io.micrometer.observation.Observation.Scope;
+import jakarta.ws.rs.client.ClientRequestContext;
+import jakarta.ws.rs.client.ClientResponseContext;
+import org.glassfish.jersey.client.spi.PostInvocationInterceptor;
+
+import static io.micrometer.jakarta.instrument.binder.http.ObservationHttpJakartaClientFilter.OBSERVATION_PROPERTY;
+import static io.micrometer.jakarta.instrument.binder.http.ObservationHttpJakartaClientFilter.OBSERVATION_SCOPE_PROPERTY;
+
+/**
+ * A post-invocation client interceptor. Will stop an observation started in
+ * {@link ObservationHttpJakartaClientFilter} in case of exceptions.
+ *
+ * @author Marcin Grzejszczak
+ * @since 1.12.0
+ */
+public class ObservationHttpJakartaInterceptor implements PostInvocationInterceptor {
```

Comment:
This should be named more specifically to Jersey, since it implements a Jersey SPI.

Reply:
I suppose this would be a problem in an environment that is using a JAX-RS implementation not including Jersey. I don't know how much of a problem that is, but as far as I understand, there will be issues with the filter linked in the JavaDoc if there are exceptions and an interceptor like this isn't configured.

Reply:
Ultimately, this is a blocker for providing JAX-RS client instrumentation. This was ran into by the OTel Java instrumentation as well in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/5430. There's more information about the limitation in JAX-RS itself in https://github.com/jakartaee/rest/issues/684. So I don't know that we can provide JAX-RS client instrumentation without it being buggy. For now, we should not merge this code until/unless we can find an alternative solution.

Reply:
@marcingrzejszczak The existing tests do not demonstrate the issue mentioned in the linked issues. Running the tests against a WireMock endpoint stubbed as follows demonstrates the issue:

```java
wmRuntimeInfo.getWireMock().register(WireMock.get("/connectionReset").willReturn(WireMock.aResponse().withFault(Fault.CONNECTION_RESET_BY_PEER)));
```

The observation will be started but never stopped in this case. The Jersey specific `ObservationJerseyClientInterceptor` doesn't have any effect in the `ObservationJaxRsHttpClientFilterTests` because RESTEasy is used. This demonstrates why I don't think we can offer JAX-RS client instrumentation: it leaks observations when there is no response from the server. Rather we can add Jersey client instrumentation to jersey-micrometer, where with the Jersey interceptor we can ensure no leaks. It's unfortunate we can't instrument at the JAX-RS level for the client side, but we don't want to put out instrumentation with a known leak.

<details>
<summary>Full test code and failure</summary>

Test method added to `ObservationJaxRsHttpClientFilterTests`
```java
@Test
void clientFilterShouldWorkWithJakartaHttpClientForExceptionsWithoutResponse(WireMockRuntimeInfo wmRuntimeInfo) {
    wmRuntimeInfo.getWireMock().register(WireMock.get("/connectionReset").willReturn(WireMock.aResponse().withFault(Fault.CONNECTION_RESET_BY_PEER)));

    try (Client client = ClientBuilder.newClient()) {
        client.register(new ObservationJaxRsHttpClientFilter(observationRegistry, null));
        final WebTarget target = client
            .target("http://localhost:" + wmRuntimeInfo.getHttpPort() + "/connectionReset");
        Exception exception = null;
        try (Response response = target.request().get()) {
            BDDAssertions.fail("Response should not be returned");
        } catch (Exception e) {
            exception = e;
        }
        BDDAssertions.then(exception).isNotNull().isInstanceOf(ProcessingException.class);
    }

    wmRuntimeInfo.getWireMock()
        .verifyThat(WireMock.getRequestedFor(WireMock.urlEqualTo("/connectionReset"))
            .withHeader("foo", WireMock.equalTo("bar")));

    TestObservationRegistryAssert.then(observationRegistry)
        .hasSingleObservationThat()
        .hasBeenStarted()
        .hasBeenStopped()
        .hasError();
}
```

Test outcome
```
java.lang.AssertionError: Observation is not stopped

	at io.micrometer.jakarta9.instrument.binder.http.jaxrs.client.ObservationJaxRsHttpClientFilterTests.clientFilterShouldWorkWithJakartaHttpClientForExceptionsWithoutResponses(ObservationJaxRsHttpClientFilterTests.java:112)
	at java.base/java.lang.reflect.Method.invoke(Method.java:578)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
```

</details>